pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,277 @@
+
+// Next checks if there is another line available in the text file.
+func (f *textReader) Next() bool {
+	return f.scanner.Scan()
+}
+
+// Scan scans the next line from the text file into the provided pointer to string.
+func (f *textReader) Scan(i interface{}) error {
+	switch target := i.(type) {
+	case *string:
+		*target = f.scanner.Text()
+		return nil
+	default:
+		return errNotPointer
+	}
+}
+
+// Close closes the FTP file connection.
+func (f *ftpFile) Close() error {
+	return f.response.Close()
+}
+
+// Name returns the name of the file.
+func (f *ftpFile) Name() string {
+	return f.name
+}
+
+// Read reads data from the FTP file into the provided byte slice and updates the file offset.
+func (f *ftpFile) Read(p []byte) (n int, err error) {
+	r, err := f.conn.RetrFrom(f.path, uint64(f.offset))
+	if err != nil {
+		f.logger.Errorf("Read Failed Failed to open file with path %q : %v", f.path, err)
+		return 0, err
+	}
+
+	n, err = r.Read(p)
+	if err != nil && !errors.Is(err, io.EOF) {
+		f.logger.Errorf("Read Failed Failed to read from %q : %v", f.path, err)
+		return 0, err
+	}
+
+	r.Close()
+
+	f.offset += int64(n)
+
+	f.logger.Logf("Read Success Read %v bytes from %q", n, f.path)
+
+	if n < len(p) {
+		return n, io.EOF
+	}
+
+	return n, nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *ftpFile) ReadAt(p []byte, off int64) (n int, err error) {
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		f.logger.Errorf("ReadAt Failed Error opening file with path %q at %v offset : %v", f.path, off, err)
+		return 0, err
+	}
+
+	n, err = resp.Read(p)
+	if err != nil && !errors.Is(err, io.EOF) {
+		f.logger.Errorf("ReadAt Failed Error reading file with path %q at %v offset : %v", f.path, off, err)
+		return 0, err
+	}
+
+	resp.Close()
+
+	f.logger.Logf("ReadAt Success Read %v bytes from %q at offset of %v", n, f.path, off)
+
+	if n < len(p) {
+		return n, io.EOF
+	}
+
+	return n, nil
+}
+
+func (f *ftpFile) check(whence int, offset, length int64) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+		if offset < 0 || offset > length {
+			return 0, ErrOutOfRange
+		}
+
+		return offset, nil
+
+	case io.SeekEnd:
+		if offset > 0 || offset < -length {
+			return 0, ErrOutOfRange
+		}
+
+		return length + offset, nil
+
+	case io.SeekCurrent:
+		if f.offset+offset >= length || f.offset+offset < 0 {
+			return 0, ErrOutOfRange
+		}
+
+		return f.offset + offset, nil
+	default:
+	}
+
+	return 0, ErrOutOfRange
+}
+
+// Seek sets the offset for the next Read or ReadAt operation.
+func (f *ftpFile) Seek(offset int64, whence int) (int64, error) {
+	n, err := f.conn.FileSize(f.path)
+	if err != nil {
+		f.logger.Errorf("Seek Failed Error : %v", err)
+		return 0, err
+	}
+
+	res, err := f.check(whence, offset, n)
+	if err != nil {
+		f.logger.Errorf("Seek Failed Error : %v", err)
+		return 0, err
+	}
+
+	f.logger.Logf("Seek Success Offset at whence %v : %v", whence, res)
+
+	return res, err
```

Comment:
The canonical way would be this, even if it's the same

```suggestion
	return res, nil
```